### PR TITLE
Modify opent2t library to be instance based & added telelmetry tracking

### PR DESCRIPTION
**Pull request by ChuckFerring-zz**
Modified the opent2t library to be instance-based.
Added telemetry tracking to logger.
Added telemetry events and exceptions to opent2t library.

This is a breaking change.  I have PR's ready to go for the other repos but this change needs to go in first.

### CODE REVIEW

File: node/lib/Logger.ts [hunk 1]
```diff
@@ -1,58 +1,55 @@
...
 
-export class Logger implements ILogger {
+export enum TraceLevel {
+    Verbose = 0,
+    Info = 1,
+    Warn = 2,
+    Error = 3
+}
+
+export class Logger {
```
Comment (arjun-msft) [inline node/lib/Logger.ts]: add a class level comment at the least #Closed

File: node/lib/Logger.ts [hunk 2]
```diff
@@ -72,15 +69,43 @@ export class Logger implements ILogger {
...
+    public exception(exception: Error, data?: { [key: string]: any; }): void {
+        this.callAllTransports(t => t.exception(exception, data));
     }
 
-    public normalizeWith(normalizer: (logObject: any) => any): ILogger {
+    public getConfiguredTransports(): Array<string> {
+        return this.logger._names;
+    }
+
+    public normalizeWith(normalizer: (logObject: any) => any): Logger {
```
Comment (arjun-msft) [inline node/lib/Logger.ts]: where is this used? #Closed

File: node/lib/OpenT2T.ts [hunk 3]
```diff
@@ -128,20 +141,23 @@ export class OpenT2T {
...
+        let startTime = Date.now();
+        this.logger.verbose(
             `getPropertyAsync for : '${propertyName}' for translator schema: ${schemaName}`);
-        let translatorForSchema = OpenT2T.getTranslatorForSchema(translator, schemaName);
-        OpenT2T.validateMemberName(propertyName);
-        let memberName = OpenT2T.uncapitalize(propertyName);
+        let trackingData: any = { propertyName, schemaName };
+        let translatorForSchema = this.getTranslatorForSchema(translator, schemaName);
+        this.validateMemberName(propertyName);
+        let memberName = this.uncapitalize(propertyName);
```
Comment (arjun-msft) [inline node/lib/OpenT2T.ts]: break up this section with some whitespace #Closed

File: node/lib/OpenT2T.ts [hunk 4]
```diff
@@ -214,6 +235,8 @@ export class OpenT2T {
             // The result object looks like a Promise. Await it to complete async the operation.
             await result;
         }
+
+        this.logger.event("SetProperty", Date.now() - startTime, trackingData);
```
Comment (arjun-msft) [inline node/lib/OpenT2T.ts:241]: this goes for all methods, you should log an event in the error cases as well #Closed

File: node/lib/OpenT2T.ts [hunk 5]
```diff
@@ -178,32 +197,34 @@ export class OpenT2T {
...
+        translator: IThingTranslator,
+        schemaName: string | ThingSchema,
+        propertyName: string,
+        value: any): Promise<void> {
+        let startTime = Date.now();
+        this.logger.verbose(
             `setPropertyAsync for : '${propertyName}' for translator schema: ${schemaName}`);
-        let translatorForSchema = OpenT2T.getTranslatorForSchema(translator, schemaName);
-        OpenT2T.validateMemberName(propertyName);
+        let trackingData: any = { propertyName, schemaName, value };
```
Comment (arjun-msft) [inline node/lib/OpenT2T.ts]: this goes for all cases where translator is passed in, log translator name #Closed

**Comment by arjun-msft** [comment]
:shipit:

File: node/lib/AppInsightsTrackerTransport.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
+import { ITrackerTransport } from "./ITrackerTransport";
+import { TraceLevel } from "./Logger";
+import * as applicationinsights from "applicationinsights";
+
+export class AppInsightsTrackerTransport implements ITrackerTransport {
+
+    private client: any;
+
+    constructor(key: string) {
+        this.client = applicationinsights.getClient(key);
```
Comment (arjun-msft) [inline node/lib/AppInsightsTrackerTransport.ts:11]: add session id #Closed

File: node/lib/Logger.ts (hunk 2 above)
Comment (ChuckFerring-zz) [reply node/lib/Logger.ts]: Doesn't look like it was used anywhere, removed.

---
In reply to: [109506842](https://github.com/openT2T/opent2t/pull/40#discussion_r109506842) [](ancestors = 109506842)

**Pull request merged**